File: java/scf-bp-component/src/main/java/org/mydotey/scf/bp/component/labeled/AbstractDcConfigurationSource.java
package org.mydotey.scf.bp.component.labeled;

import java.util.Collection;
import java.util.Objects;

import org.mydotey.scf.ConfigurationSourceConfig;
import org.mydotey.scf.labeled.AbstractLabeledConfigurationSource;
import org.mydotey.scf.labeled.PropertyLabel;

/**
 * @author koqizhao
 *
 * May 17, 2018
 */
public abstract class AbstractDcConfigurationSource
        extends AbstractLabeledConfigurationSource<ConfigurationSourceConfig> {

    public AbstractDcConfigurationSource(ConfigurationSourceConfig config) {
        super(config);
    }

    @Override
    protected Object getPropertyValue(Object key, Collection<PropertyLabel> labels) {
        if (key.getClass() != String.class)
            return null;

        String app = null;
        String dc = null;
        if (labels != null) {
            for (PropertyLabel l : labels) {
                if (Objects.equals(l.getKey(), DcSetting.APP_KEY))
                    app = (String) l.getValue();

                if (Objects.equals(l.getKey(), DcSetting.DC_KEY))
                    dc = (String) l.getValue();
            }
        }

        return getPropertyValue((String) key, app, dc);
    }

    protected String getPropertyValue(String key, String app, String dc) {
        DcSetting dcSetting = getPropertyValue(new DcSetting(key, app, dc, null));
        return dcSetting == null ? null : dcSetting.getValue();
    }

    protected abstract DcSetting getPropertyValue(DcSetting setting);

}

File: dotnet/src/MyDotey.SCF.BP.Component/Labeled/AbstractDcConfigurationSource.cs
using System;
using System.Collections.Generic;

using MyDotey.SCF.Labeled;

namespace MyDotey.SCF.BP.Component.Labeled
{
    /**
     * @author koqizhao
     *
     * May 17, 2018
     */
    public abstract class AbstractDcConfigurationSource
        : AbstractLabeledConfigurationSource<ConfigurationSourceConfig>
    {

        public AbstractDcConfigurationSource(ConfigurationSourceConfig config)
            : base(config)
        {

        }

        protected override object GetPropertyValue(object key, ICollection<IPropertyLabel> labels)
        {
            if (!(key is string))
                return null;

            String app = null;
            String dc = null;
            if (labels != null)
            {
                foreach (IPropertyLabel l in labels)
                {
                    if (object.Equals(l.Key, DcSetting.APP_KEY))
                        app = (String)l.Value;

                    if (object.Equals(l.Key, DcSetting.DC_KEY))
                        dc = (String)l.Value;
                }
            }

            return GetPropertyValue((String)key, app, dc);
        }

        protected virtual String GetPropertyValue(String key, String app, String dc)
        {
            DcSetting dcSetting = GetPropertyValue(new DcSetting(key, app, dc, null));
            return dcSetting == null ? null : dcSetting.Value;
        }

        protected abstract DcSetting GetPropertyValue(DcSetting setting);
    }
}